Utils/Dataflow/009_oracleConnector/Oracle2JSON.py

```diff
@@ -0,0 +1,307 @@
+    """
+    a single-pass iterator (for a generator) restructuring list of datasets:
+    from view:
+    [
+       {"taskid": 1, "type": "input", "datasetname": "First_in"},
+       {"taskid": 1, "type": "input", "datasetname": "Second_in"},
+       {"taskid": 1, "type": "input", "datasetname": "Third_in"},
+       {"taskid": 1, "type": "output", "datasetname": "First_out"},
+       {"taskid": 1, "type": "output", "datasetname": "Second_out"},
+       {"taskid": 1, "type": "output", "datasetname": "Third_out"},
+       {"taskid": 2, "type": "input", "datasetname": "First_in"},
+       {"taskid": 2, "type": "output", "datasetname": "First_out"},
+       {"taskid": 2, "type": "output", "datasetname": "Second_out"},
+    ...]
+     to
+    [
+       {"taskid": 1,
+        "input": ["First_in", "Second_in", "Third_in"],
+        "output": ["First_out", "Second_out", "Third_out"]
+       },
+       {"taskid": 2,
+        "input": ["First_in"],
+        "output": ["First_out","Second_out"]
+       },
+    ...]
+    """
+    grouper = defaultdict(lambda: defaultdict(list))
+    for d in rec:
+        grouper[d['taskid']][d['type']].append(d['datasetname'])
+    return [dict(taskid=k, **v) for k, v in grouper.items()]
+
+
+def join_results(tasks, datasets):
+    d = defaultdict(dict)
+    buffers = (tasks, datasets)
+    join_buffer = {}
+    req_n = len(buffers)
+    for l in buffers:
+        for elem in l:
+            tid = elem['taskid']
+            d[tid].update(elem)
+            n = join_buffer.get(tid, 0) + 1
+            if n == req_n:
+                yield d[tid]
+                del join_buffer[tid]
+                del d[tid]
+            else:
+                join_buffer[tid] = n
+
+    # Flush records stuck in the buffer (not joined)
+    for tid in d:
+        yield d[tid]
+
+
+def process(conn, offset_date, final_date_cfg, step_seconds, queries):
+    if final_date_cfg:
+        final_date = final_date_cfg
+    else:
+        final_date = date2str(datetime.now())
+    break_loop = False
+    while (not break_loop and str2date(offset_date) < str2date(final_date)):
+        end_date = date2str(str2date(offset_date)
+                            + timedelta(seconds=step_seconds))
+        if str2date(end_date) > str2date(final_date):
+            end_date = final_date
+            break_loop = True
+        sys.stderr.write("(TRACE) %s: Run queries for interval from %s to %s\n"
+                         % (date2str(datetime.now()), offset_date, end_date))
+        if mode == SQUASH_POLICY:
+            records = squash(conn, queries, offset_date, end_date)
+        elif mode == PLAIN_POLICY:
+            records = plain(conn, queries, offset_date, end_date)
+        for r in records:
+            r['phys_category'] = get_category(r)
+            sys.stdout.write(json.dumps(r) + '\n')
+        update_offset(end_date)
+        offset_date = end_date
+        if not final_date_cfg:
+            final_date = date2str(datetime.now())
```

Comment:
final_date is determined at 133-136. Should it be redetermined at every loop iteration?

Reply:
Yes.
For the initial data load the loop can interate for quite a long time.
If we asked the script to load all the data from `$then` till `$now`, we want to have this `$now` up to date -- however long it took to reach it.